Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceDomGenerator.cpp

```diff
@@ -178,18 +178,19 @@ namespace AzToolsFramework
             PrefabDomPath relativeDomPathFromTopToEntity(relativePathFromTop.c_str());
 
             // Retrieves the entity DOM from the template of the focused or root instance.
-            PrefabDom focusedOrRootTemplateDom;
-            focusedOrRootTemplateDom.CopyFrom(m_prefabSystemComponentInterface->FindTemplateDom(focusedOrRootInstance->GetTemplateId()),
+            PrefabDom focusedOrRootTemplateDomCopy;
+            focusedOrRootTemplateDomCopy.CopyFrom(
+                m_prefabSystemComponentInterface->FindTemplateDom(focusedOrRootInstance->GetTemplateId()),
                 entityDom.GetAllocator());
```

Comment:
You should use focusedOrRootTemplateDomCopy.GetAllocator() here, not entityDom.GetAllocator().

The allocator of the document being copied from is template-inferred from the type of the first parameter.  But the second parameter is which allocator to use for NEW allocations made in the NEW document its going to create.  This means you should always pass in the allocator that belongs to the document you are copying into.  If you use the other document's allocaotr, then the memory will be freed and become invalid when that other document goes out of scope.

Reply:
Interesting. These changes are coming from the original prototype code. I am worried if other places have something similar.

So the call stack is:
```
PrefabPublicHandler::SomeFunction()
  - InstanceDomGenerator::GenerateInstanceDom()
    - InstanceDomGenerator::UpdateContainerEntityInDomFromRoot()
```
Currently, the allocator defined and owned in the top level (SomeFunction) is getting passed down and used in subsequent calls. So, from what you said, for this use case it won't cause any issue because the allocator won't be invalidated.

BUT it doesn't look like a good pattern since it might fail in other cases. So, the better solution would be:

```cpp
PrefabDom focusedOrRootTemplateDomCopy;
focusedOrRootTemplateDomCopy.CopyFrom(m_prefabSystemComponentInterface->FindTemplateDom(focusedOrRootInstance->GetTemplateId()), 
    focusedOrRootTemplateDomCopy.GetAllocator());

// Changes the DOM and makes copies using focusedOrRootTemplateDomCopy's allocator

// At the end of the function
entityDom.CopyFrom(<dom_we_what_to_copy_from>, entityDom.GetAllocator());
```

This makes sure that we always use the appropriate allocator in current scope.

Reply:
yes.  Its a bit weird for us to passing our alloocators around in PrefabDom anyway since its supposed to mostly hide the fact that there's a rapidJSON (As opposed to a different kind of dom) under there, but allocators being passed around is a rapidjson specific feature.

also, as far as I can tell,  the default bool that CopyFrom takes (copyconststrings) only applies to strings that are in a pre-allocated memory block when in-situ parsing is used. You can give rapidjson a string to Parse() and you can tell it to use that string as its memory for const strings.  (In situ-parsing).  If you do that, when it creates strings, it sets their pointer to offsets in the given block of memory, and sets a flag on them that indicates that its allocator does not own them.    That default bool flag only applies to that specific case.  In all other cases, strings are always copied by allocating more from the allocator.